contracts/DsaWrapper.sol

```diff
@@ -1,62 +1,244 @@
+        external
+        payable
+        onlyAuthority(_id, msg.sender)
+    {
+        // get dsa-address
         address dsaAddress = instapool.getAccountIdDetails(_id).account;
 
-        address[] memory accountAuthorities = instapool.getAccountAuthorities(dsaAddress);
+        string[] memory targets = new string[](1);
+        targets[0] = "BASIC-A";
 
-        return accountAuthorities;
+        bytes[] memory data = new bytes[](1);
+        bytes4 basicDeposit = bytes4(
+            keccak256("deposit(address,uint256,uint256,uint256)")
+        );
+
+        address tokenAddress = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
+
+        data[0] = abi.encodeWithSelector(
+            basicDeposit,
+            tokenAddress,
+            msg.value,
+            0,
+            0
+        );
+
+        IDSA(dsaAddress).cast{value: msg.value}(targets, data, address(this));
     }
 
-    // deposit ether to dsa
-    function depositEtherToDsa(uint256 _id) external payable {
-        // get dsa-address
+    /**
+     * @dev withdraws ETH from DSA account given by _id
+     * @param _id id of DSA account
+     * @param _amt amount of ETH to withdraw
+     */
+    function withdrawEther(uint256 _id, uint256 _amt)
+        external
+        onlyAuthority(_id, msg.sender)
+    {
         address dsaAddress = instapool.getAccountIdDetails(_id).account;
+        require(dsaAddress.balance >= _amt, "INSUFFICIENT FUNDS");
 
         string[] memory targets = new string[](1);
         targets[0] = "BASIC-A";
 
         bytes[] memory data = new bytes[](1);
-        bytes4 basicDeposit = bytes4(keccak256("deposit(address,uint256,uint256,uint256)"));
+        bytes4 basicWithdraw = bytes4(
+            keccak256("withdraw(address,uint256,address,uint256,uint256)")
+        );
 
         address tokenAddress = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
 
-        data[0] = abi.encodeWithSelector(basicDeposit, tokenAddress, msg.value, 0, 0);
+        data[0] = abi.encodeWithSelector(
+            basicWithdraw,
+            tokenAddress,
+            _amt,
+            msg.sender,
+            0,
+            0
+        );
 
-        console.log(abc.isAuth(msg.sender)); // false
-        console.log(abc.isAuth(address(this))); // false
+        IDSA(dsaAddress).cast(targets, data, address(this));
+    }
+
+    /**
+     * @dev deposits ERC20 token to DSA account given by _id
+     * @param _id id of DSA account
+     * @param _amt amount of tokens to be deposited
+     * @param _tokenAddress address of token to be deposited
+     */
+    function depositErc20(
+        uint256 _id,
+        uint256 _amt,
+        address _tokenAddress
+    ) external payable onlyAuthority(_id, msg.sender) {
+        IERC20 token = IERC20(_tokenAddress);
 
-        IDSA(dsaAddress).cast{value: msg.value}(targets, data, address(0));
+        // get dsa-address
+        address dsaAddress = instapool.getAccountIdDetails(_id).account;
+
```

Comment:
Let's also check whether `dsa` has enough `amt` or not.

Reply:
This has to be checked in withdrawErc20 right? I've already placed a check for that...